### PR TITLE
ci: dualstack overlay windows test cases

### PR DESCRIPTION
**Pull request by paulyufan2**
**Reason for Change**:

dualstack overlay windows test cases
dualstack overlay windwos conformance test

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 1]
```diff
@@ -115,3 +117,41 @@ steps:
...
+    name: "WindowsDualStackOverlayControlPlaneScaleTests"
+    displayName: "Windows DualStack Overlay ControlPlane Scale Tests"
+    retryCountOnTaskFailure: 3
+
+  - script: |
+      echo "DualStack Overlay DataPath Test"
+      cd test/integration/datapath
+      sudo -E env "PATH=$PATH" go test -count=1 datapath_windows_test.go -timeout 3m -tags connection -restartKubeproxy true -run ^TestDatapathWin$
+    name: "WindowsDualStackOverlayDatapathTests"
+    displayName: "Windows DauStack Overlay Datapath Tests"
```
Comment (ecigar13) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:156]: Typo.

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 2]
```diff
@@ -65,6 +65,8 @@ steps:
       sudo go test -count=1 datapath_linux_test.go -timeout 3m -tags connection -run ^TestDatapathLinux$ -tags=connection,integration -isDualStack=true
       echo "Dualstack Overlay Linux datapath IPv4 test"
       sudo go test -count=1 datapath_linux_test.go -timeout 3m -tags connection -run ^TestDatapathLinux$ -tags=connection,integration
+      echo "cleaning up load-test namespace"
```
Comment (jpayne3506) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:68]: Should this get changed to use make test-load now?

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 3]
```diff
@@ -115,3 +117,52 @@ steps:
...
+  - task: AzureCLI@2
+    inputs:
+      azureSubscription: $(BUILD_VALIDATIONS_SERVICE_CONNECTION)
+      scriptLocation: "inlineScript"
+      scriptType: "bash"
+      addSpnToEnvironment: true
+      inlineScript: |
+        set -e
+        make -C ./hack/aks windows-nodepool-up AZCLI=az SUB=$(SUB_AZURE_NETWORK_AGENT_BUILD_VALIDATIONS) CLUSTER=${{ parameters.clusterName }}
+        echo "Windows node are successfully added to DualStack Overlay Cluster"
```
Comment (jpayne3506) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:141]: nit: past tense. 
```suggestion
        echo "Windows node have been successfully added to DualStack Overlay Cluster"
```

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 4]
```diff
@@ -115,3 +117,52 @@ steps:
...
+        kubectl cluster-info
+        kubectl get node -owide
+        kubectl get po -owide -A
+    name: "Add_Windows_Node"
+    displayName: "Add windows node on dualstack overlay cluster"
```
Comment (jpayne3506) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:146]: nit: can we change the `displayName` to be shorter? 
```suggestion
    name: "Add_Windows_Node"
    displayName: "Add Windows Nodepool"
```

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 5]
```diff
@@ -115,3 +117,52 @@ steps:
...
+    displayName: "Windows DualStack Overlay ControlPlane Scale Tests"
+    retryCountOnTaskFailure: 3
+
+  - script: |
+      echo "DualStack Overlay DataPath Test"
+      cd test/integration/datapath
+      sudo -E env "PATH=$PATH" go test -count=1 datapath_windows_test.go -timeout 3m -tags connection -restartKubeproxy true -run ^TestDatapathWin$
+    name: "WindowsDualStackOverlayDatapathTests"
+    displayName: "Windows DaulStack Overlay Datapath Tests"
+    retryCountOnTaskFailure: 3
```
Comment (jpayne3506) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:168]: Should we add a Node restart for the windows nodepool as well? This needs to be potentially added to https://github.com/Azure/azure-container-networking/blob/9200af8aa1a45d6fca281b32494cadf3a22f935d/.pipelines/singletenancy/azure-cni-overlay/azure-cni-overlay-e2e-step-template.yaml#L190 as well

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml (hunk 2 above)
Comment (pjohnst5) [reply .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:68]: This is a different test suite, this is under the test/integration/datapath directory, `make test-load` tests the files under `test/integration/load`, so I think we should leave it as is

**Review by pjohnst5 — APPROVED** [review]
LGTM

**Review by jpayne3506 — CHANGES_REQUESTED** [review]

**Comment by pjohnst5** [comment]
If you wanted to get past the error you see in the pipeline (failed to overwrite existing config in kubeconfig), you could add the `--overwrite-existing` in `hack/aks/Makefile`
https://learn.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-get-credentials()

But you would have to commit a change anyways, which would re-trigger a new build

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml (hunk 2 above)
Comment (jpayne3506) [reply .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:68]: This is also in reference to the CNS `make test-integration` and restartCNS scenario that existed in the file before.

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 6]
```diff
@@ -115,3 +117,52 @@ steps:
...
+        make -C ./hack/aks windows-nodepool-up AZCLI=az SUB=$(SUB_AZURE_NETWORK_AGENT_BUILD_VALIDATIONS) CLUSTER=${{ parameters.clusterName }}
+        echo "Windows node are successfully added to DualStack Overlay Cluster"
+        kubectl cluster-info
+        kubectl get node -owide
+        kubectl get po -owide -A
+    name: "Add_Windows_Node"
+    displayName: "Add windows node on dualstack overlay cluster"
+
+  - script: |
+      nodeList=`kubectl get node -owide | grep Windows | awk '{print $1}'`
```
Comment (vipul-21) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:149]: nit: we can get the nodes based on the label `os=windows`

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml [hunk 7]
```diff
@@ -115,3 +117,52 @@ steps:
...
+  - ${{ if eq( parameters['testDropgz'], true) }}:
+      - script: |
+          echo "##vso[task.setvariable variable=dropgzVersion]$(make cni-dropgz-test-version)"
+        name: dropgzTestVersion
+        displayName: "Dropgz Test Version"
+  - ${{ else }}:
+      - script: |
+          echo "##vso[task.setvariable variable=dropgzVersion]$(make cni-dropgz-version)"
+        name: dropgzVersion
+        displayName: "Dropgz Version"
```
Comment (vipul-21) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:130]: These should be moved above line 39 and also be changed in task `integrationTest` to use the dropgz veriosn variable.  Currently the pipeline can test different dropgz/cni in these tasks based on the parameter(test dropgz) in a single run.

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml (hunk 5 above)
Comment (paulyufan2) [reply .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:168]: yes, I will use another PR for both v4 and dualstack windows node restart

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml (hunk 6 above)
Comment (paulyufan2) [reply .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-step-template.yaml:149]: can you please explain more about this?

File: .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml [hunk 8]
```diff
@@ -6,8 +6,8 @@ parameters:
 
 
 jobs:
-  - job: cni_k8se2e
-    displayName: "CNI k8s E2E"
+  - job: CNI_${{ parameters.os }}
+    displayName: CNI_k8s_E2E_${{ parameters.os }}
```
Comment (jpayne3506) [inline .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml:10]: Can we remove the `_` for the displayName?

File: .pipelines/cni/k8s-e2e/k8s-e2e-step-template.yaml [hunk 9]
```diff
@@ -28,6 +28,12 @@ steps:
         kubectl taint nodes -l kubernetes.azure.com/mode=system node-role.kubernetes.io/control-plane:NoSchedule
       fi
 
+      # Taint Windows nodes so that windows tests do not run on them
```
Comment (jpayne3506) [inline .pipelines/cni/k8s-e2e/k8s-e2e-step-template.yaml:31]:
```suggestion
      # Taint Windows nodes so that Linux tests do not run on them
```

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-job-template.yaml [hunk 10]
```diff
@@ -61,10 +61,22 @@ stages:
...
           portforward: true
           hostport: true
           service: true
+
+      - template: ../../cni/k8s-e2e/k8s-e2e-job-template.yaml
+        parameters:
+          sub: $(BUILD_VALIDATIONS_SERVICE_CONNECTION)
+          clusterName: ${{ parameters.clusterName }}-$(commitID)
+          os: windows
+          dependsOn: cni_linux
```
Comment (jpayne3506) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-job-template.yaml:77]: Is this working as intended even though the job is `CNI_${{ parameters.os }}`?

File: .pipelines/cni/k8s-e2e/k8s-e2e-step-template.yaml (hunk 9 above)
Comment (jpayne3506) [reply .pipelines/cni/k8s-e2e/k8s-e2e-step-template.yaml:31]: Nit

File: .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml [hunk 11]
```diff
@@ -6,8 +6,8 @@ parameters:
 
 
 jobs:
-  - job: cni_k8se2e
-    displayName: "CNI k8s E2E"
+  - job: CNI_${{ parameters.os }}
```
Comment (pjohnst5) [inline .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml:9]: I think this change, as well as the change in `.pipelines/cni/k8s-e2e/k8s-e2e-step-template.yaml`, should be in a separate PR

They don't seem related to dualstack overlay windows testing

File: .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-job-template.yaml [hunk 12]
```diff
@@ -61,10 +61,22 @@ stages:
         parameters:
           sub: $(BUILD_VALIDATIONS_SERVICE_CONNECTION)
           clusterName: ${{ parameters.clusterName }}-$(commitID)
-          os: ${{ parameters.os }}
+          os: linux
```
Comment (pjohnst5) [inline .pipelines/singletenancy/dualstack-overlay/dualstackoverlay-e2e-job-template.yaml:64]: I see in master branch, we only pass linux as an OS
I think we should not hard code it here to linux, but leave it as parameter
https://github.com/Azure/azure-container-networking/blob/e874736dba7b9d0e2ffbbf7cbcf2a1241555f070/.pipelines/pipeline.yaml#L455

File: .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml (hunk 11 above)
Comment (pjohnst5) [reply .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml:9]: Hmm, just saw you add this into the dualstack step
Is there a reason why?

**Review by pjohnst5 — CHANGES_REQUESTED** [review]
Just want to remove that hardcoding of "`linux`", and question about new k8s step

File: .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml (hunk 11 above)
Comment (paulyufan2) [reply .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml:9]: it will run windows conformance tests for dualstack

Comment (jpayne3506) [reply .pipelines/cni/k8s-e2e/k8s-e2e-job-template.yaml:9]: This is because there are 2 calls of this template in a single stage. This requires a the `job:` field to be unique for multiple calls and not possible without giving it parameter

**Review by vipul-21 — APPROVED** [review]

**Review by jpayne3506 — APPROVED** [review]

**Comment by rbtr** [comment]
@paulyufan2 pls don't dismiss someone else's review requesting changes, you don't know if you have sufficiently addressed their comments (or even if you do, it's not clear to those of us reading what happened here, now) and it is [up to the reviewer to resolve those](https://github.com/Azure/azure-container-networking/blob/master/docs/contributing/github.md?plain=1#L20) and approve
cc @pjohnst5
